File: SList.cpp
#include "SList.h"
#include <iostream>
#include <fstream>
#include <vector>
#include <stdexcept>

using namespace std;


void SList::printReverse(SLNode* node) const {
    if (node == nullptr) return;
    printReverse(node->next);
    cout << "\"" << node->data << "\"";
    if (node != nullptr) cout << " <- ";
}

SList::SList() {
    head = nullptr;
    size = 0;
}

SList::~SList() {
    SLCLEAR();
}

void SList::SLCLEAR() {
    SLNode* current = head;
    while (current != nullptr) {
        SLNode* temp = current;
        current = current->next;
        delete temp;
    }
    head = nullptr;
    size = 0;
    cout << "Список очищен." << endl;
}

void SList::SLPUSH_HEAD(const string& value) {
    SLNode* newNode = new SLNode;
    newNode->data = value;
    newNode->next = head;
    head = newNode;
    size++;
    //cout << "Элемент \"" << value << "\" добавлен в голову списка" << endl;
}

void SList::SLPUSH_TAIL(const string& value) {
    SLNode* newNode = new SLNode;
    newNode->data = value;
    newNode->next = nullptr;
    
    if (head == nullptr) {
        head = newNode;
    } else {
        SLNode* current = head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = newNode;
    }
    size++;
    //cout << "Элемент \"" << value << "\" добавлен в хвост списка" << endl;
}

void SList::SLPUSH_BEFORE(const string& target, const string& value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    if (head->data == target) {
        SLPUSH_HEAD(value);
        return;
    }
    
    SLNode* current = head;
    while (current->next != nullptr && current->next->data != target) {
        current = current->next;
    }
    
    if (current->next != nullptr) {
        SLNode* newNode = new SLNode;
        newNode->data = value;
        newNode->next = current->next;
        current->next = newNode;
        size++;
        //cout << "Элемент \"" << value << "\" добавлен перед \"" << target << "\"" << endl;
    }
    else {
        //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
    }
}

void SList::SLPUSH_AFTER(const string& target, const string& value) {
    SLNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }
    
    if (current != nullptr) {
        SLNode* newNode = new SLNode;
        newNode->data = value;
        newNode->next = current->next;
        current->next = newNode;
        size++;
        //cout << "Элемент \"" << value << "\" добавлен после \"" << target << "\"" << endl;
    }else{

            //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
    }
}

void SList::SLREMOVE_HEAD() {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    SLNode* temp = head;
    head = head->next;
    //cout << "Элемент \"" << temp->data << "\" удалён из головы списка" << endl;
    delete temp;
    size--;
}

void SList::SLREMOVE_TAIL() {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    if (head->next == nullptr) {
        //cout << "Элемент \"" << head->data << "\" удалён из хвоста списка" << endl;
        delete head;
        head = nullptr;
    }
    else {
        SLNode* current = head;
        while (current->next->next != nullptr) {
            current = current->next;
        }
        //cout << "Элемент \"" << current->next->data << "\" удалён из хвоста списка" << endl;
        delete current->next;
        current->next = nullptr;
    }
    size--;
}

void SList::SLREMOVE_BEFORE(const string& target) {
    if (head == nullptr || head->next == nullptr) {
        cout << "Недостаточно элементов для удаления!" << endl;
        return;
    }
    
    if (head->data == target) {
        cout << "Невозможно удалить элемент перед головой!" << endl;
        return;
    }
    
    if (head->next->data == target) {
        SLREMOVE_HEAD();
        return;
    }
    
    SLNode* current = head;
    while (current->next->next != nullptr && current->next->next->data != target) {
        current = current->next;
    }
    
    if (current->next->next != nullptr) {
        SLNode* temp = current->next;
        current->next = current->next->next;
        //cout << "Элемент \"" << temp->data << "\" удалён перед \"" << target << "\"" << endl;
        delete temp;
        size--;
    }
    else {
        cout << "Элемент \"" << target << "\" не найден или перед ним нет элемента" << endl;
    }
}

void SList::SLREMOVE_AFTER(const string& target) {
    SLNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }
    
    if (current != nullptr && current->next != nullptr) {
        SLNode* temp = current->next;
        current->next = current->next->next;
        //cout << "Элемент \"" << temp->data << "\" удалён после \"" << target << "\"" << endl;
        delete temp;
        size--;
    }
    else {
        cout << "Элемент \"" << target << "\" не найден или после него нет элемента" << endl;
    }
}

void SList::SLREMOVE_VALUE(const string& value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    if (head->data == value) {
        SLNode* temp = head;
        head = head->next;
        //cout << "Элемент \"" << temp->data << "\" (голова) удалён по значению" << endl;
        delete temp;
        size--;
        return;
    }
    
    SLNode* current = head;
    while (current->next != nullptr && current->next->data != value) {
        current = current->next;
    }
    
    if (current->next != nullptr) {
        SLNode* temp = current->next;
        current->next = current->next->next;
        //cout << "Элемент \"" << temp->data << "\" удалён по значению" << endl;
        delete temp;
        size--;
    }
    else {
        //cout << "Элемент \"" << value << "\" не найден в списке" << endl;
    }
}

bool SList::SLSEARCH(const string& value) const {
    SLNode* current = head;
    size_t index = 0;
    while (current != nullptr) {
        if (current->data == value) {
            //cout << "Элемент \"" << value << "\" найден на позиции " << index << endl;
            return true;
        }
        current = current->next;
        index++;
    }
    //cout << "Элемент \"" << value << "\" не найден в списке" << endl;
    return false;
}

string SList::SLGET(size_t index) const {
    if (index >= size) {
        throw out_of_range("Индекс " + to_string(index) + " вне диапазона [0, " + to_string(size - 1) + "]");
    }
    
    SLNode* current = head;
    for (size_t i = 0; i < index; i++) {
        current = current->next;
    }
    return current->data;
}

void SList::SLPRINT_FORWARD() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    cout << "Список [" << size << "] (прямой порядок): ";
    SLNode* current = head;
    while (current != nullptr) {
        cout << "\"" << current->data << "\"";
        if (current->next != nullptr) {
            cout << " -> ";
        }
        current = current->next;
    }
    cout << " -> NULL" << endl;
}

void SList::SLPRINT_BACKWARD() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    cout << "Список [" << size << "] (обратный порядок): NULL <- ";
    printReverse(head);
    cout << endl;
}

void SList::SLPRINT_HEAD() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
    }
    else {
        cout << "Голова списка: \"" << head->data << "\"" << endl;
    }
}

void SList::SLPRINT_TAIL() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    
    SLNode* current = head;
    while (current->next != nullptr) {
        current = current->next;
    }
    cout << "Хвост списка: \"" << current->data << "\"" << endl;
}

size_t SList::SLLENGTH() const {
    return size;
}

File: slist.go
package main

import (
	"fmt"
)

type SLNode struct {
	data string
	next *SLNode
}

type SList struct {
	head *SLNode
	size int
}

func NewSList() *SList {
	return &SList{
		head: nil,
		size: 0,
	}
}

func (l *SList) SLCLEAR() {
	l.head = nil
	l.size = 0
	fmt.Println("Список очищен.")
}

func (l *SList) SLPUSH_HEAD(value string) {
	newNode := &SLNode{
		data: value,
		next: l.head,
	}
	l.head = newNode
	l.size++
}

func (l *SList) SLPUSH_TAIL(value string) {
	newNode := &SLNode{
		data: value,
		next: nil,
	}

	if l.head == nil {
		l.head = newNode
	} else {
		current := l.head
		for current.next != nil {
			current = current.next
		}
		current.next = newNode
	}
	l.size++
}

func (l *SList) SLPUSH_BEFORE(target, value string) {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	if l.head.data == target {
		l.SLPUSH_HEAD(value)
		return
	}

	current := l.head
	for current.next != nil && current.next.data != target {
		current = current.next
	}

	if current.next != nil {
		newNode := &SLNode{
			data: value,
			next: current.next,
		}
		current.next = newNode
		l.size++
	}
}

func (l *SList) SLPUSH_AFTER(target, value string) {
	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current != nil {
		newNode := &SLNode{
			data: value,
			next: current.next,
		}
		current.next = newNode
		l.size++
	}
}

func (l *SList) SLREMOVE_HEAD() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	l.head = l.head.next
	l.size--
}

func (l *SList) SLREMOVE_TAIL() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	if l.head.next == nil {
		l.head = nil
	} else {
		current := l.head
		for current.next.next != nil {
			current = current.next
		}
		current.next = nil
	}
	l.size--
}

func (l *SList) SLREMOVE_BEFORE(target string) {
	if l.head == nil || l.head.next == nil {
		fmt.Println("Недостаточно элементов для удаления!")
		return
	}

	if l.head.data == target {
		fmt.Println("Невозможно удалить элемент перед головой!")
		return
	}

	if l.head.next.data == target {
		l.SLREMOVE_HEAD()
		return
	}

	current := l.head
	for current.next.next != nil && current.next.next.data != target {
		current = current.next
	}

	if current.next.next != nil {
		current.next = current.next.next
		l.size--
	}
}

func (l *SList) SLREMOVE_AFTER(target string) {
	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current != nil && current.next != nil {
		current.next = current.next.next
		l.size--
	}
}

func (l *SList) SLREMOVE_VALUE(value string) {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	if l.head.data == value {
		l.SLREMOVE_HEAD()
		return
	}

	current := l.head
	for current.next != nil && current.next.data != value {
		current = current.next
	}

	if current.next != nil {
		current.next = current.next.next
		l.size--
	}
}

func (l *SList) SLSEARCH(value string) bool {
	current := l.head
	index := 0
	for current != nil {
		if current.data == value {
			return true
		}
		current = current.next
		index++
	}
	return false
}

func (l *SList) SLGET(index int) (string, error) {
	if index >= l.size || index < 0 {
		return "", fmt.Errorf("индекс %d вне диапазона [0, %d]", index, l.size-1)
	}

	current := l.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	return current.data, nil
}

func (l *SList) SLPRINT_FORWARD() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	fmt.Printf("Список [%d] (прямой порядок): ", l.size)
	current := l.head
	for current != nil {
		fmt.Printf("\"%s\"", current.data)
		if current.next != nil {
			fmt.Print(" -> ")
		}
		current = current.next
	}
	fmt.Println(" -> NULL")
}

func (l *SList) printReverse(node *SLNode) {
	if node == nil {
		return
	}
	l.printReverse(node.next)
	fmt.Printf("\"%s\"", node.data)
	if node != l.head {
		fmt.Print(" <- ")
	}
}

func (l *SList) SLPRINT_BACKWARD() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	fmt.Printf("Список [%d] (обратный порядок): NULL <- ", l.size)
	l.printReverse(l.head)
	fmt.Println()
}

func (l *SList) SLLENGTH() int {
	return l.size
}

// Для сериализации
func (l *SList) GetAll() []string {
	result := make([]string, 0, l.size)
	current := l.head
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	return result
}

func (l *SList) SetAll(items []string) {
	l.SLCLEAR()
	for i := len(items) - 1; i >= 0; i-- {
		l.SLPUSH_HEAD(items[i])
	}
}